packages/wallet/ui/src/bridge-dapp.jsx

```diff
@@ -0,0 +1,180 @@
+// @ts-check
+import '@endo/eventual-send/shim';
+import './lockdown.js';
+
+import React, { useEffect } from 'react';
+import ReactDOM from 'react-dom';
+import { BridgeProtocol } from '@agoric/web-components';
+import { addOffer } from './store/Offers.js';
+import { loadDapp, upsertDapp, watchDapps } from './store/Dapps.js';
+
+Error.stackTraceLimit = Infinity;
+
+/**
+ * Sends a message to the dapp in the parent window.
+ *
+ * @param {*} payload
+ */
+const sendMessage = payload => window.parent.postMessage(payload, '*');
+
+/**
+ * Ensures this bridge is being loaded inside another window.
+ */
+const checkParentWindow = () => {
+  const me = window;
+  const { parent } = window;
+  if (me === parent) {
+    throw Error('window.parent === parent!!!');
+  }
+};
+
+/**
+ * Requests a dapp at a given origin to connect to the wallet UI. This will
+ * prompt the user to accept the dapp in the wallet and allow it to propose
+ * offers if accepted.
+ *
+ * @param {string} origin
+ * @param {string} chainId
+ * @param {string} address
+ * @param {string} proposedPetname
+ */
+const requestDappConnection = (origin, chainId, address, proposedPetname) => {
+  const dapp = loadDapp(chainId, address, origin);
+  if (dapp) {
+    return;
+  }
+  upsertDapp(chainId, address, { origin, petname: proposedPetname });
+};
+
+/**
+ * Watches for changes in local storage to the dapp's approval status and
+ * notifies the dapp.
+ *
+ * @param {string} origin
+ * @param {string} chainId
+ * @param {string} address
+ * @param {boolean} latestValue
+ */
+const watchDappApproval = (origin, chainId, address, latestValue) => {
+  watchDapps(chainId, address, dapps => {
+    console.log('something happened!!!', dapps);
+    const dapp = dapps.find(d => d.origin === origin);
+    const isDappApproved = dapp && dapp.isEnabled;
+    if (isDappApproved !== latestValue) {
+      sendMessage({
+        type: BridgeProtocol.dappApprovalChanged,
+        isDappApproved,
+      });
+      latestValue = isDappApproved;
+    }
+  });
+};
+
+const createAndAddOffer = (origin, chainId, address, config) => {
+  const currentTime = new Date().getTime();
+  const id = currentTime;
```

Comment:
this isn't reliable, due to different clocks. please `lastOfferId` from wallet `.current`. If you deem that out of scope, please put a TODO here. Bonus points for a ticket :)

Reply:
> We cannot reference wallet .current here without completely reverting this implementation

We don't need to reference `.current` here. Just include `id`  in the `config` param.

Please do that and if you must use time when making the config, 1) make it use UTC time so it doesn't break when people move time zones and 2) make a TODO to use `.current`.

Reply:
I think its better for the wallet to control the IDs instead of relying on the dapp to provide something unique. The dapp also doesn't currently load the offers and that would be more work and computation which seems unnecessary. 

Doesn't this just have to be a unique identifier? If we don't trust the milliseconds to not collide between two offers, maybe a more sufficiently randomness function? I don't see how different clocks/timezones could break this because the milliseconds would still have to match which is improbable.

Reply:
> Doesn't this just have to be a unique identifier?

No, it has to be higher than all previous ids. Each id sets a new high-water mark.

> I think its better for the wallet to control the IDs instead of relying on the dapp to provide something unique.

Is your argument that the dapp needn't know the state of the wallet? I see the case for that when considering `dapp-psm` proposing a swap, but since multiple agents can be changing wallet balances I do think dapps need to be aware of the wallet `.current` state.

Reply:
Closing the loop: we agree fixing this is out of scope for this bridge work because it's an existing problem and will be fixed by https://github.com/Agoric/agoric-sdk/issues/6478